### DISCUSSION TITLE
Pivotal Tuning CLI and how to use them.

### DISCUSSION THREAD

**Discussion by cloneofsimo** [cloneofsimo/lora#121] [Announcements]
![image](https://user-images.githubusercontent.com/35953539/211209093-247e9d72-e93c-48bd-a173-558129ef36b8.png)

Most of the recent updates are about lora_pti : CLI for Pivotal Tuning Inversion, with various tricks and techs to get extreme performance LoRA training output.

All of the README examples, including example above, were built with `lora_pti` CLI with [DEFAULT PARAMETERS](https://github.com/cloneofsimo/lora/blob/master/training_scripts/multivector_example.sh) Here are new parameters and what they mean:

## Extended Latent + Dataset

```
  --placeholder_tokens="<s1>|<s2>" \ 
  --placeholder_token_at_data="<tok>|<s1><s2>"\
```

So, with recent extended latent, you can train multiple tokens for textual inversion, and they are declared with placeholder_tokens argument `"<TOK1>|<TOK2>|..."`, each of them seperated with `|` character.
If you are using caption dataset (That is, file name of the dataset is caption), you might want to map certain token in caption to the token of your need.

For example, let's say you initialized two tokens `<s1>|<s2>`, and you have an image file : ` a photo of <tok> holding flowers .jpg` , you want to substitute `<tok>` with `<s1> wearing <s2>`. Then, you use argument : 

```bash
  --placeholder_token_at_data="<tok>|<s1> wearing <s2>"\
```

The argument will transform the caption to :  `a photo of <s1> wearing <s2> holding flowers ` for the image.

Also, if you want to use template caption, use the argument : 

```
 --use_template="style" \ Use this if you have no caption as the filename
```

There is type `style` and `object` available.

## Mask Conditioned Training

```
 --use_face_segmentation_condition
```

Use this if you want to focus on faces. Else, remove this line. Check [this](https://github.com/cloneofsimo/lora/discussions/96) if you are interested what this does. 

## Training steps for Two stages

So there is two stage in PTI. One is Bayesian training textual inversion with high learning rate, and one is training LoRA.

```
  --max_train_steps_ti=1000 \ Number of steps for stage 1.
  --max_train_steps_tuning=1000 \ Number of steps for stage 2. 1000 is more than enough.
```

If the concept is difficult, you want stage 1 steps to be higher. But 1000 is very likely to be ok. In many cases, having 500 ti, 500 tuning works just as ok. 
If you are going to have smaller learning rate, you would definitely want to bump up these values, as it would need longer training time.

## Other arguments (default is probably OK)

Here are kind of unimportant ones / or ones you might be familiar with. You can use the default values, but if you want to know : 

```
  --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5"  \  Base model to fine tune.
  --instance_data_dir="./data/some_data" \ Directory where the dataset exists.
  --output_dir="./model/lora1" \ Directory where output loras will be saved 
  --train_text_encoder \ This will train the text encoder as well.
  --resolution=512 \ 
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --scale_lr \ This will re-scale learning rate. You probably want this on.
  --initializer_tokens="st|at" \ If you are using n tokens, you want to have n initializer tokens, seperated with | . 
  --learning_rate_unet=1e-4 \ Unet learning rate. this works ok.
  --learning_rate_text=1e-5 \ Text encoder learning rate. this works ok.
  --learning_rate_ti=5e-4 \ Textual inversion learning rate. This works ok.
  --color_jitter \ This will augment dataset's color slightly. Usually ok to turn on unless your color matters like insane.
  --lr_scheduler="linear" \ This will be a scheduler for textual inversion only. LoRA will be added in the future.
  --lr_warmup_steps=0 \ Warmup steps for textual inversion learning rate
  --save_steps=100 \ Saving frequency. Keep this small. LoRA is tiny anyways
  --perform_inversion=True \ You want this unless you just want to train LoRA.
  --clip_ti_decay \ This will keep the norm of the textual inversion intact. You definitely want this on.
  --weight_decay_ti=0.000 \ This will even further make textual inversion smaller. You don't need this positive if you have clip_ti_decay option on.
  --weight_decay_lora=0.001\ 
  --continue_inversion \ Whether to continue training embedding after stage 1.  
  --continue_inversion_lr=1e-4 \ If you do, then with what learning rate? You probably want this small.
  --device="cuda:0" \ Device to train on
    --lora_rank=4 \ Rank of the LoRA
```

If you have any questions, comment below, I will update this discussion whenever there is an update on the CLI part.

**Comment by abi** [comment]
Thanks for your amazing work here! Excited to follow along the progress.

Can you explain what initializer_tokens and placeholder_tokens mean? I'm looking to train a particular person's face, where should I put a name/unique token associated with them and where the class token ("man", "girl") go?

**Reply by cloneofsimo** [reply]
They are both from textual inversion concepts.

Placeholder tokens are just like the ones from textual inversion. You will have to use these tokens in a prompt to get a result later. 

`a photo of <s1><s2> holding a cup`

Initializer tokens are yes like a class token, but many of them to match the number of placeholder tokens.

**Reply by geocine** [reply]
How about `--initializer_tokens="girl|<rand>|<rand>"` What does `<rand>` mean?

**Reply by geocine** [reply]
```
--placeholder_tokens="<s1>|<s2>|<s3>" \
--placeholder_token_at_data="<s>|<s1><s2><s3>"\
--initializer_tokens="girl|<rand>|<rand>" \
```

I am looking at this example , if this was the setting. What would the prompt look like?

**Reply by cloneofsimo** [reply]
@geocine  That is depreciated. Use https://github.com/cloneofsimo/lora/blob/master/training_scripts/multivector_example.sh this instead. Now, token `<rand-X>` will initialize with gaussian random $N(0, X^2)$ . I recommend using `<rand-0.018>`, so in above case,

```
--placeholder_tokens="<s1>|<s2>|<s3>" \
--placeholder_token_at_data="<s>|<s1><s2><s3>"\
--initializer_tokens="girl|<rand-0.018>|<rand-0.018>" \
```

**Reply by geocine** [reply]
Okay, what would my prompt look like after training the model. I understand `<s1>` will be replace with `girl` but what about `<s2>` and `<s3>` if its rand? Will `photo of a girl` show my trained model?

**Reply by cloneofsimo** [reply]
It is just how it will be initialized, and will become a meaningful token after training. You will have to do 
`photo of a <s1><s2><s3>` to get a result.

**Reply by FrancescoSaverioZuppichini** [reply]
I don't understand, I `n` picture of a person's face, how should I call the images? What should I use in the CI? Can somebody give a real example and explain why?

**Reply by harrywang** [reply]
@cloneofsimo thanks for the great work. After reading the comments here, I still don't understand the meaning of `<s1>` and `<s2>` - should there just be one special token if I train for one style using text inversion? why s1 and s2? I am very confused. It would be great if you can give any pointers. Thanks a lot!

**Reply by guangyangsjc18** [reply]
I also do not understand. If you have multiple tokens, shouldn't you have multiple instance_data_dir as well? Or these are inferred by the subfolder names ?

**Reply by strangehelix** [reply]
Reading about textual inversions may help https://huggingface.co/docs/diffusers/training/text_inversion

**Comment by FunWithFaces** [comment]
Thanks for this!

Can you let us know the minimum VRAM that is required to run this?

Also -- I'm wondering if it might be possible to provide an example with public domain images that we could use to try to reproduce a test result and make sure everything is set up properly. I appreciate that this probably is not a priority for you at the moment!

**Reply by cloneofsimo** [reply]
I think you need at least 15GB for this. I'm not definitely sure, because nvidia-smi doesn't tell you everything. Can someone confirm on this?

**Reply by geocine** [reply]
I'm on windows right now mine is using around ~17GB

**Reply by cloneofsimo** [reply]
Right. So PTI isn't using any hardware optimizations, so you might need more VRAM than other methods. I'll work on it sooner or later.

**Reply by cloneofsimo** [reply]
With gradient checkpoint + xformer, 12GB might be enough, but can someone confirm?
![image](https://user-images.githubusercontent.com/35953539/211401659-6910ff18-6336-4be9-b62c-6cfdf3df3bd7.png)

**Reply by 2kpr** [reply]
Working on my 3080 Ti at about 11.9/12.2GB

**Reply by FunWithFaces** [reply]
Does --use_xformers work with the lora_pti CLI, or is there something more I need to do to get this going? I'm still getting OOM with 12GB, seems like I must be close!

**Reply by FunWithFaces** [reply]
Edit -- don't do this, see cloneofsimo comment below

> In case anyone else is wondering about this, I added xformers support by modifying cli_lora_pti.py by adding the line
> 
> from lora_diffusion.xformers_utils import set_use_memory_efficient_attention_xformers
> 
> after the "from lora_diffusion import" block (line 35)
> 
> and then added:
> 
>     set_use_memory_efficient_attention_xformers(unet, True)
>     set_use_memory_efficient_attention_xformers(vae, True)
> 
> after the gradient checkpointing command on line ~600

worked a charm. This was just based on what i saw in the older training scripts. Probably a better way to go about it I have no idea what I'm really doing

**Reply by cloneofsimo** [reply]
I added it in the lora_pti,I think it should work by now with `--enable_xformers_memory_efficient_attention` argument

**Reply by FunWithFaces** [reply]
OK so, it looks like there is exactly one xformers version that functions with this right now, I'm just conveying this info from the helpful folks in the auto1111 dreambooth extension discord.

if on linux, there is a wheel available via
https://github.com/Zuxier/xformers/releases/download/torch1.13-0.14dev0-python3.10/xformers-0.0.14.dev0-cp310-cp310-linux_x86_64.whl

if you want to attempt to build yourself it is 
git clone https://github.com/facebookresearch/xformers.git
cd xformers
git checkout e23b369c09468

also note that another (not sure github handle, insignia | Real Deal on discord) pointed out that you may have issues with the face segmentation stuff without

changing:
[Image.open(f) for f in self.instance_images_path]

to :
[Image.open(f).convert('RGB') for f in self.instance_images_path]
(in around line 188 i think, dataset.py)

I am still new to github and havent figured out how to submit PR yet, I'll get there

**Comment by geocine** [comment]
@cloneofsimo . This is my output running your example at https://github.com/cloneofsimo/lora/blob/master/training_scripts/use_face_conditioning_example.sh

![image](https://user-images.githubusercontent.com/507464/211357962-ebd6345b-e1d9-4255-a43b-46f6be55f23f.png)

1. What is the purpose of the `_inv_` files?
2. I want to merge these to an existing `ckpt` so I need to convert these `safetensors` to `pt` then use `lora_add`? If so, how do I convert these `safetensors` to `pt`. I only see `pt_to_safetensors` but not the other way around. Also I understand `lora_add` only works with `pt` . Need your guide how to proceed.

**Reply by cloneofsimo** [reply]
1. _inv files are basically stage 1 outputs : textual inversion files. you can use them just like any other safetensors format.
2. You can use them like the following, with `patch_pipe` function.
In 0.1.0, you can use both .pt and .safetensors for `patch_pipe`.

![image](https://user-images.githubusercontent.com/35953539/211363800-533cfe45-2984-456e-b344-0375ac3091c4.png)

**Reply by geocine** [reply]
Thanks, I tried this and it works. Still want to know how to merge this to an existing model so I can use it on Auto's webui.

**Reply by cloneofsimo** [reply]
Nice to see that it works for you. Does it has the level of quality you want? I (and the community) would love/appreciate some feedbacks

**Reply by Elevory** [reply]
The `patch_pipe` script works for me as well, and the quality is good. However, the lora_add function seems broken. Cannot get it to merge even after converting to ckpt format. For what it's worth, I'm using SD 2.1 as a base. Inference wasn't working on 1.5.

lora_add result with safetensors:

```
  File "T:\programs\anaconda3\envs\diffusers\lib\site-packages\torch\serialization.py", line 1002, in _legacy_load
    magic_number = pickle_module.load(f, **pickle_load_args)
_pickle.UnpicklingError: could not find MARK
```

lora_add result with ckpt:

```
  File "T:\programs\anaconda3\envs\diffusers\lib\site-packages\lora_diffusion\lora.py", line 419, in weight_apply_lora
    up_weight = loras.pop(0).detach().to(weight.device)
KeyError: 0
```

**Reply by cloneofsimo** [reply]
Huh, ok. I'll add this to an issue.

**Reply by geocine** [reply]
@Elevory what did you use to convert the `safetensors` to `pt`? 

@cloneofsimo I just tried the settings on face segmentation and I'm getting good results at 1 out of 5. I haven't added negatives though. Is there a way to explicitly use the vae  vae-ft-mse-840000-ema-pruned.ckpt on hhe inference?

**Reply by SU1199** [reply]
@geocine 
Can you share your pip freeze.. I've been trying since 7 hours to replicate this to no avail :(

**Reply by geocine** [reply]
Sure I'll get back to you in 2 hours

**Reply by Elevory** [reply]
> @Elevory what did you use to convert the `safetensors` to `pt`?

I used this: https://github.com/diStyApps/Safe-and-Stable-Ckpt2Safetensors-Conversion-Tool-GUI

**Reply by geocine** [reply]
Thanks @Elevory  , @SU1199 , I'm using a conda environment here you go https://gist.github.com/geocine/5f4523725dbd8b7343de28f37fc45720

**Reply by DTG777** [reply]
I'm having some trouble running this file. Could you please tell me how you run this file to train the lora model？

**Comment by GabrielTomasel** [comment]
using civitai various models and others that i trained myself i get "Rank should be the same per model" error using the cloneofsimo/lora from replicate. What can i do to make the LoRAs compatible with that repo? Using Automatic1111 those loras work ok.
i see that --lora_rank=4 \ Rank of the LoRA is an option but dont know what that means of how can i set that in Kohya colab notebook to train my models.

**Comment by weidaru** [comment]
All of the examples use `<s1>|<s2>` as placeholder tokens. Any reason behind that? If I train a style, is a single token `<s1>` good enough?

**Comment by Lucky-Light-Sun** [comment]
hello, I have a question about the word "CLI". I know LoRA, and I know a little about Pivotal Tuning. But what does the word "CLI" mean? Is this short for CLIP? Or it refers to a new paper?

**Reply by shaokun** [reply]
CLI is command line interface...

**Reply by Lucky-Light-Sun** [reply]
Got it, thank you

**Comment by TeaCult** [comment]
I really dont understand how to prepare data. There are some xxx.json some xxx.yaml  other datasets have imagename.png and imagename.txt, I cannot relate <tok> <s1> <s2> etc to these. I could not find a clear guide to explain this. Peft is different , your lora and automaitic 1111 lora dataset seems different. Civit AI has an article to scape and train lora it is different. Hugging face diffusers lora is also different. Can you please explain this for a confused person :). 

I am ready to start from textual inversion and most basic. But since guides and codes have api like middlewares I cannot understand what lies behind and its logic.

Thank you.

**Reply by atacanpolat** [reply]
I'd like to know this as well :)

**Comment by TeaCult** [comment]
Also I put safetensors to automatic 1111 seems that it does not work. Either I could not get <s1><s2> dataset tags right or safetensors does not work with it. In that case, How can I convert safetensors to pt ? btw believe me "I have googled a lot"
